onnxruntime/core/providers/cpu/tensor/upsample_aa.h

```diff
@@ -0,0 +1,682 @@
+     */
+    if (x < 0.0f) {
+      x = -x;
+    }
+    if (x < 1.0f) {
+      return ((cubic_coeff_a + 2.0f) * x - (cubic_coeff_a + 3.0f)) * x * x + 1;
+    }
+    if (x < 2.0f) {
+      return (((x - 5.0f) * x + 8.f) * x - 4.f) * cubic_coeff_a;
+    }
+    return 0.0f;
+  }
+};
+
+struct TriLinearParamsAA : FilterParamsAA {
+  float filter(float x) const override {
+    if (x < 0.0f) {
+      x = -x;
+    }
+    if (x < 1.0f) {
+      return 1.0f - x;
+    }
+    return 0.0f;
+  }
+};
+
+template <typename T>
+struct AccumulateType {
+  using type = int32_t;
+  using Dtype = T;
+};
+
+template <>
+struct AccumulateType<int32_t> {
+  using type = float;
+};
+
+template <>
+struct AccumulateType<float> {
+  using type = float;
+};
+
+template <>
+struct AccumulateType<double> {
+  using type = double;
+};
+
+void SetupUpsampleFilterAA(FilterParamsAA& p,
+                           const gsl::span<int64_t> input_h_w_c,
+                           const gsl::span<int64_t> output_h_w_c,
+                           const gsl::span<float> scale_h_w_c,
+                           const std::vector<float>& roi,
+                           AllocatorPtr& alloc,
+                           const GetOriginalCoordinateFunc& get_original_coordinate,
+                           const int32_t dtype, bool exclude_outside, const bool is_nchw);
+template <class T>
+inline constexpr bool is_8bit_v = std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value;
+
+template <typename T>
+void UpsampleBaseAA(FilterParamsAA& p,
+                    const int64_t batch_size,
+                    const int64_t num_channels,
+                    const int64_t input_height,
+                    const int64_t input_width,
+                    const int64_t output_height,
+                    const int64_t output_width,
+                    const bool use_extrapolation,
+                    const float extrapolation_value,
+                    const T* const XdataBase,
+                    T* const YdataBase,
+                    AllocatorPtr& alloc,
+                    concurrency::ThreadPool* tp) {
+  const uint8_t* clip8_lookups = &p.clip8_lookups_table[640];
+
+  auto image_temp_buffer = BufferUniquePtr(alloc->Alloc(static_cast<size_t>(input_height *
+                                                                            output_width * num_channels) *
+                                                        sizeof(T)),
+                                           BufferDeleter(alloc));
+
+  using ACtype = typename AccumulateType<T>::type;
+
+  for (int64_t n = 0; n < batch_size; ++n) {
+    auto* temp_buffer = static_cast<T*>(image_temp_buffer.get());
+    // horizon interpolate
+
+    concurrency::ThreadPool::TrySimpleParallelFor(
+        tp, narrow<std::ptrdiff_t>(num_channels),
+        [&](std::ptrdiff_t c) {
+          const T* const Xdata =
+              XdataBase + (n * num_channels + (c)) *
+                              (input_height * input_width);
```

Comment:
nit: this could all be a single line within the 120 char limit from the look of it.